include/deal.II/lac/trilinos_tpetra_precondition.h

```diff
@@ -0,0 +1,1311 @@
+       * @param dst Input vector to apply the preconditioner to
+       * @param src Result vector
+       */
+      virtual void
+      vmult(Vector<Number, MemorySpace>       &dst,
+            const Vector<Number, MemorySpace> &src) const;
+
+      /**
+       * @brief Apply the transpose preconditioner
+       *
+       * @param dst Input vector to apply the preconditioner to
+       * @param src Result vector
+       */
+      virtual void
+      Tvmult(Vector<Number, MemorySpace>       &dst,
+             const Vector<Number, MemorySpace> &src) const;
+
+      /**
+       * @brief Apply the preconditioner
+       *
+       * @param dst Input vector to apply the preconditioner to
+       * @param src Result vector
+       */
+      virtual void
+      vmult(dealii::Vector<Number> &dst, dealii::Vector<Number> &src) const;
+
+      /**
+       * @brief Apply the transpose preconditioner
+       */
+      virtual void
+      Tvmult(dealii::Vector<Number> &dst, dealii::Vector<Number> &src) const;
+
+      /**
+       * @brief Access to underlying Trilinos data
+       *
+       * Calling this function from an uninitialized object will cause an
+       * exception.
+       */
+      Teuchos::RCP<TpetraTypes::LinearOperator<Number, MemorySpace>>
+      trilinos_operator() const;
+
+      /**
+       * @name Partitioners
+       */
+      /** @{ */
+
+      /**
+       * Return the partitioning of the domain space of this matrix, i.e., the
+       * partitioning of the vectors this matrix has to be multiplied with.
+       *
+       * @return IndexSet of the domain of this preconditioner
+       */
+      IndexSet
+      locally_owned_domain_indices() const;
+
+      /**
+       * Return the partitioning of the range space of this matrix, i.e., the
+       * partitioning of the vectors that are result from matrix-vector
+       * products.
+       *
+       * @return IndexSet of the range of this preconditioner
+       */
+      IndexSet
+      locally_owned_range_indices() const;
+
+      /** @} */
+
+      /**
+       * @addtogroup Exceptions
+       */
+      /** @{ */
+      /**
+       * The maps of the underlying matrix and one of the vectors
+       * do not match.
+       */
+      DeclException1(
+        ExcNonMatchingMaps,
+        std::string,
+        << "The sparse matrix the preconditioner is based on "
+        << "uses a map that is not compatible to the one in vector " << arg1
+        << ". Check preconditioner and matrix setup.");
+
+      /**
+       * The chosen preconditioner does not support a transposed apply.
+       */
+      DeclExceptionMsg(
+        ExcTransposeNotSupported,
+        "The chosen preconditioner does not support transposing the matrix.");
+      /** @} */
+
+    protected:
+      /**
+       * This is a pointer to the preconditioner object that is used when
+       * applying the preconditioner.
+       */
+      Teuchos::RCP<TpetraTypes::LinearOperator<Number, MemorySpace>>
+        preconditioner;
```

Comment:
That's neat with this base type, which covers all kinds of methods in the `vmult` operations without further intervention.

Reply:
Thanks :) It certainly made adding all these preconditioners simpler. 
Hardest parts where finding and understanding the changes between parameters 
and distilling that into a documentation which is hopefully better to understand.